src/Pate/Memory/MemTrace.hs

```diff
@@ -500,22 +506,43 @@ instance TestEquality (SymExpr sym) => Eq (MemOp sym ptrW) where
      , valR == valR'
      , Just Refl <- testEquality valO valO'
     = cond == cond' && dir == dir' && end == end'
-  MergeOps p opsT opsF == MergeOps p' opsT' opsF'
-    | Just Refl <- testEquality p p'
-    = opsT == opsT' && opsF == opsF'
+
   _ == _ = False
 
+instance OrdF (SymExpr sym) => Ord (MemOp sym ptrW) where
+  compare (MemOp (LLVMPointer reg1 off1) dir1 cond1 sz1 (LLVMPointer vr1 vo1) end1)
+          (MemOp (LLVMPointer reg2 off2) dir2 cond2 sz2 (LLVMPointer vr2 vo2) end2) =
+    case compareF sz1 sz2 of
+      LTF -> LT
+      GTF -> GT
+      EQF ->
+        (compare reg1 reg2) <>
+        (toOrdering $ compareF off1 off2) <>
+        compare dir1 dir2 <>
+        (compare cond1 cond2) <>
+        (compare vr1 vr2) <>
+        (toOrdering $ compareF vo1 vo2) <>
+        compare end1 end2
+
+data MemEvent sym ptrW where
+  MemOpEvent :: MemOp sym ptrW -> MemEvent sym ptrW
+  SyscallEvent :: forall sym ptrW w.
+    -- TODO, something more realistic
+    SymExpr sym (BaseBVType w) ->
+    MemEvent sym ptrW
```

Comment:
What do you think about making a separate module to define observable events? I suspect we might want to generate them in places besides the memory model (especially when we start adding support for e.g., calls to external functions)